### PR TITLE
Enable GLM recurrent boundary restore with DCP and DFlash

### PR DESCRIPTION
**Pull request by voipmonitor**
## Purpose

Enable GLM-5.3-Flash semantic recurrent checkpoints with static DFlash2 and
decode context parallelism. Shared leading instructions, completed prompts,
and committed response endpoints can be reused without retaining a recurrent
state every 256 tokens.

## Restore contract

- Attention snapshots use each cache group's context-parallel geometry.
  Target pages are sharded under DCP; DFlash draft pages remain replicated.
- DFlash restores its projected context KV and generates the first proposal
  through its ordinary query graph. It must not rewrite the last context row
  into an immutable shared page.
- Evicted draft pages are allowed only before the sliding window required by
  the checkpoint. A hole inside that window invalidates the checkpoint.
- The existing collective worker wait completes all rank-local copies before
  the scheduler publishes the checkpoint. No additional handshake is introduced.

External KV connectors, dynamic speculation, and unsupported parallel layouts
continue to select aligned retention. This PR does not implement semantic
checkpoint persistence in LMCache or change physical serving-page defaults.

## Dependencies and scope

The serving qualification stack includes #674 for scalar-safe recurrent
restoration and #664 for isolated logits-only scheduler steps. It also includes
#676, which fixes partial physical attention-page lookup independently.

This does not duplicate those changes: it supplies the missing DCP address
mapping and DFlash context/window restore contract. Open-PR searches in this
fork and upstream found no equivalent request-boundary implementation.

## Validation

Status: implemented; GPU-resident cache correctness is qualified for the
configurations below. Matched release-performance qualification is in progress.

Four RTX PRO 6000 Blackwell Workstation Edition GPUs, TP4, stock clocks,
4,096-token scheduler budget, 2,048-token target/recurrent pages, and full plus
piecewise CUDA graphs:

- 42 CUDA attention snapshot cases cover DCP1/2/4, replicated draft groups,
  logical page edges, exact copied bytes, and untouched neighboring pages.
- 18 CUDA scalar-restore cases pass with #674.
- 260 scheduler/prefix-cache/DFlash CPU cases, 128 prefix-manager cases, and
  12 partial-attention-rewind cases pass.
- DFlash2/DCP1 and DFlash2/DCP4 with NVFP4 target KV restore all 8,191,
  8,192, 8,193, and 32,768 prompt tokens and preserve all 256 greedy output
  token IDs. Response continuations match independent cold controls.
- MTP3/DCP4 passes the exact, instruction, response, and eight-conversation
  matrix. Five seeded 4,096-token coding requests show no acceptance collapse
  after restoration: median accepted draft tokens/step is 2.095 cold and
  2.157 restored; verifier throughput is 92.32 and 92.45 steps/s.
- No-speculation/DCP4 with FP8 target KV passes 32 packaged-image scenarios,
  including actual assistant/user continuation, tool history, and divergent
  instructions. The assistant continuation restores 11,385 tokens; the tool
  history replays all 11,520 prompt tokens.
- DFlash2/DCP4 with NVFP4 target KV passes the same 32-scenario packaged-image
  matrix, including the first restored proposal through the query CUDA graph.
- 18 CPU configuration cases verify automatic GLM no-spec/MTP/DFlash boundary
  selection under DCP1/2/4 and fail-closed external-cache selection. The complete
  prefix-primitive test file passes all 46 cases.
- `llama-benchy` 0.4.0 at depth 8,192 passes coherence and prefix reuse with
  DFlash2/DCP1, DFlash2/DCP4, and MTP3/DCP4.

Representative reproduction commands, in a built vLLM environment:

```bash
uv run .venv/bin/python -m pytest -q \
  tests/v1/worker/test_boundary_checkpoint_attention.py \
  tests/v1/spec_decode/test_dflash_dcp.py \
  tests/v1/core/prefix_cache/test_partial_prefix_cache_primitives.py \
  tests/v1/core/prefix_cache/test_partial_prefix_cache_hits.py

uvx llama-benchy --base-url http://127.0.0.1:5051/v1 \
  --model GLM-5.3-Flash-NVFP4 --depth 8192 --enable-prefix-caching --runs 3
```

Serving probes use `/v1/completions` with token-ID prompts, `temperature=0`,
`max_tokens=256`, `ignore_eos=true`, `return_token_ids=true`, and an unchanged
`cache_salt`. Each exact prompt is sent three times. The two replays must report
the entire prompt as cached and return the same output token IDs as the first
request. Appended-response probes are compared with separately salted cold
controls. Chat probes check literal record values through instruction branches,
assistant continuations, and tool histories.

Matched release-performance and external-cache qualification remain required
before recommending the image for deployment. The stochastic acceptance medians
above are not a speedup claim.

AI assistance was used for implementation, review, and validation. Human
maintainer review is required before merge.

## Summary by CodeRabbit

- **New Features**
  - Added support for DFlash speculative decoding with GLM5 models.
  - Enabled decode context parallelism across multiple workers for GLM5.
  - Improved reuse of request-boundary checkpoints across distributed configurations.

- **Bug Fixes**
  - Improved checkpoint handling for sliding-window attention after pages are evicted.
  - Preserved correct context and attention state reuse during DFlash decoding.
  - Improved isolation when reusing partial checkpoints.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

Please check the rank-local versus replicated attention-page mapping, immutable DFlash context handling, sliding-window reachability, and external-cache fail-closed gate. Packaged TP4/DCP4 no-spec FP8 and DFlash2 NVFP4 serving passed the exact/instruction/assistant/tool/concurrency matrix. Performance and LMCache release qualification are separate and still running.

**Comment by voipmonitor** [comment]
Additional no-spec/DCP1 qualification found a limitation of the greedy-output
oracle, not a reproducible failure specific to this PR.

TP4, four RTX PRO 6000 Blackwell Workstation GPUs at stock clocks, FP8 target
KV, 4,096-token scheduler budget, 2,048-token pages, `request_boundaries`:

- Exact prompt replays, literal document answers, shared instructions, real
  assistant/tool history, eight concurrent conversations, and endpoint-hit
  counters pass.
- An appended periodic-text continuation sometimes differs from a separately
  salted full-prefill control: `The` versus ` The` at a sentence boundary.
- Ten identical **cold** controls, all reporting zero prefix hits, produce two
  distinct greedy output sequences. This happens both in the composed stack
  containing this PR and in **R27 + #674/#676 without this PR**.
- The control without this PR also differs between a cold calculation and its
  exact replay. Thus a single cold generation is not a bit-stable reference
  for this input. The kernel-level cause has not been established.

The reproducible input is token pattern
`[785,3974,13867,38627,34041,916,279,15666,5562,13]`, repeated/truncated to
8,191 tokens. Generate 256 tokens greedily three times under one cache salt;
append those output IDs and the first five pattern IDs; generate another 256
tokens. Compare with ten fresh cache salts, requesting `logprobs=10` and
`return_token_ids=true`. Use `ignore_eos=true` throughout.

No numerical kernel was changed to make this probe pass. Strict failures are
retained, not relabeled as passing. The byte-copy tests and literal-answer /
source-attribution checks remain independent correctness evidence. The
previously reported DFlash and DCP4 results retain their stated scope.
